lib/charms/opensearch/v0/opensearch_plugin_manager.py

```diff
@@ -303,32 +330,85 @@ def _compute_settings(
+        keystore_ready = True
+        cluster_settings_changed = False
+        try:
+            # If security is not yet initialized, this code will throw an exception
+            self._keystore.delete(config.secret_entries_to_del)
+            self._keystore.add(config.secret_entries_to_add)
+            if config.secret_entries_to_del or config.secret_entries_to_add:
+                self._keystore.reload_keystore()
+        except (OpenSearchKeystoreNotReadyYetError, OpenSearchHttpError):
+            # We've failed to set the keystore, we need to rerun this method later
+            # Postpone the exception now and set the remaining config.
+            keystore_ready = False
 
         current_settings, new_conf = self._compute_settings(config)
-        if current_settings == new_conf:
-            # Nothing to do here
-            logger.info("apply_config: nothing to do, return")
-            return False
+        if current_settings and new_conf and current_settings != new_conf:
+            if config.config_entries_to_del:
+                # Clean to-be-deleted entries
+                self._opensearch.request(
+                    "PUT",
+                    "/_cluster/settings?flat_settings=true",
+                    payload={"persistent": {key: "null" for key in config.config_entries_to_del}},
+                )
+                cluster_settings_changed = True
+            if config.config_entries_to_add:
+                # Configuration changed detected, apply it
+                self._opensearch.request(
+                    "PUT",
+                    "/_cluster/settings?flat_settings=true",
+                    payload={"persistent": config.config_entries_to_add},
+                )
+                cluster_settings_changed = True
 
-        # Update the configuration
+        # Update the configuration files
         if config.config_entries_to_del:
             self._opensearch_config.delete_plugin(config.config_entries_to_del)
         if config.config_entries_to_add:
             self._opensearch_config.add_plugin(config.config_entries_to_add)
-        return True
+
+        if cluster_settings_changed:
+            # We have changed the cluster settings, clean up the cache
+            del self.cluster_config
+
+        if not keystore_ready:
+            # We need to rerun this method later
+            raise OpenSearchKeystoreNotReadyYetError()
+
+        # Final conclusion, we return a restart is needed if:
+        # (1) configuration changes are needed and applied in the files; and (2)
+        # the node is not up. For (2), we already checked if the node was up on
+        # _cluster_settings and, if not, cluster_settings_changed=True.
+        return all(
+            [
+                (config.secret_entries_to_add or config.secret_entries_to_del),
+                (
+                    (config.config_entries_to_add or config.config_entries_to_del)
+                    and not cluster_settings_changed
```

Comment:
Maybe I'm missing something, but aren't these 2 conditions exclusive? 
`cluster_settings_changed == True` effectively means from the previous logic that `(config.config_entries_to_add or config.config_entries_to_del)`

Reply:
Not necessarily. We may have one case (now, more an exception), where:
1) we have a change (i.e. config_entries_to_{add,del} are not None)
2) the cluster is active, but we did not use the API calls to apply the configs from (1)

This is the only situation that we must restart, as the node must be restart so the changes take place.

Reply:
That means:
```
        # Update the configuration files
        if config.config_entries_to_del:
            self._opensearch_config.delete_plugin(config.config_entries_to_del)
        if config.config_entries_to_add:
            self._opensearch_config.add_plugin(config.config_entries_to_add)
```

may do a change, but `cluster_settings_changed = False`.

Reply:
However, that expression I wrote at the end is indeed very convoluted. I will simplify that.